### PR TITLE
Add specification, configs, new transaction type and header fields

### LINKED ISSUE #4862: Add specification and configs
Resolves #4862

Before new tx type and block header field can be used, specation should be added.

## Changes:
- Add block based fork for EIP-4844
- Add current block and EIP-4884 flag across the application
- Introduce configs to run outside and inside of interop docker setup, as part of devnet

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by flcl42**
Depends on #4850

EIP-4844 transaction type, fork rules and configs for devnet have been added anticipating transaction serialization, extended validation and processing functionality

## Changes:
- Update transaction structure to hold blob transaction and its network form
- Add configs
- Add header field
- Add spec

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Serialization.Rlp/HeaderDecoder.cs [hunk 1]
```diff
@@ -156,6 +162,11 @@ public class HeaderDecoder : IRlpValueDecoder<BlockHeader>, IRlpStreamDecoder<Bl
                 blockHeader.BaseFeePerGas = rlpStream.DecodeUInt256();
             }
 
+            if (blockHeader.Timestamp >= Eip4844TransitionTimestamp)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Serialization.Rlp/HeaderDecoder.cs:165]: quick question, do we have the spec for both Withdrawals + 4844? we're adding field in withdrawals too

File: src/Nethermind/Nethermind.Serialization.Rlp/HeaderDecoder.cs [hunk 2]
```diff
@@ -207,6 +218,11 @@ public void Encode(RlpStream rlpStream, BlockHeader? header, RlpBehaviors rlpBeh
             {
                 rlpStream.Encode(header.BaseFeePerGas);
             }
+
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Serialization.Rlp/HeaderDecoder.cs:221]: could you add roundtrip decoding/encoding tests?

**Review by MarekM25 — COMMENTED** [review]
check these tests: `Eip_transitions_loaded_correctly(),` @smartprogrammer93 is working on them too

File: src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs [hunk 3]
```diff
@@ -118,5 +122,10 @@ private bool ValidateSignature(Signature? signature, IReleaseSpec spec)
 
             return !spec.ValidateChainId || (signature.V == 27 || signature.V == 28);
         }
+
+        private bool Validate4844Fields(Transaction transaction, IReleaseSpec spec)
+        {
+            return transaction.Type != TxType.Blob || spec.IsEip4844Enabled ^ transaction.MaxFeePerDataGas is null;
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs:128]: hmmm are you sure that this condition is correct? adding TxValidatorTests might help you

File: src/Nethermind/Nethermind.Core/BlockHeader.cs [hunk 4]
```diff
@@ -47,6 +49,20 @@ public class BlockHeader
             GasLimit = gasLimit;
             Timestamp = timestamp;
             ExtraData = extraData;
+            ExcessDataGas = excessDataGas;
+            ParentExcessDataGas = parentExcessDataGas;
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Core/BlockHeader.cs:53]: do we need ParentExcessDataGas here? Is ParentHash not enough? It sounds to be the similar situation to EIP1559. Please, check BaseFeeCalculator.cs class

File: src/Nethermind/Nethermind.Core/Transaction.cs [hunk 5]
```diff
@@ -57,6 +58,13 @@ public class Transaction
         public UInt256 Timestamp { get; set; }
 
         public AccessList? AccessList { get; set; } // eip2930
+        public UInt256? MaxFeePerDataGas { get; set; } // eip4844
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Core/Transaction.cs:61]: What about changes in TxDecoder? separate PR?

File: src/Nethermind/Nethermind.Runner/Nethermind.Runner.csproj [hunk 6]
```diff
@@ -101,6 +101,9 @@
     <None Remove="out\**" />
     <None Remove="keystore\**" />
   </ItemGroup>
+  <ItemGroup>
+    <None Remove="configs\eip4844_docker.cfg" />
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Runner/Nethermind.Runner.csproj:105]: Cosmetic: to remove?

File: src/Nethermind/Nethermind.Serialization.Rlp/HeaderDecoder.cs (hunk 1 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Serialization.Rlp/HeaderDecoder.cs:165]: Yep, the current PR waits for @smartprogrammer93 changes

File: src/Nethermind/Nethermind.Runner/Nethermind.Runner.csproj (hunk 6 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Runner/Nethermind.Runner.csproj:105]: Yes, let me remove it. It will be needed to participate in the devnet, but let it be just there, outside of our repo

File: src/Nethermind/Nethermind.Core/Transaction.cs (hunk 5 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Core/Transaction.cs:61]: Yes, all the encoding part will be a separate RP

File: src/Nethermind/Nethermind.Core/BlockHeader.cs (hunk 4 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Core/BlockHeader.cs:53]: I'll remove it now and return in the gas calculations request, if it will be still needed. It's hard to show for now why it's useful

File: src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs [hunk 7]
```diff
@@ -85,6 +88,7 @@ private bool ValidateChainId(Transaction transaction)
                     return true;
                 case TxType.AccessList:
                 case TxType.EIP1559:
+                case TxType.Blob:
```
[the remark this answers is not in the document]
Comment (flcl42) [reply src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs:91]: I found one more test and some interesting `UserOperationTxBuilder` thing, which I made a note to check later. Do you have any other ideas?

File: src/Nethermind/Nethermind.Specs/MainNetSpecProvider.cs [hunk 8]
```diff
@@ -57,11 +58,13 @@ public void UpdateMergeTransitionInfo(long? blockNumber, UInt256? terminalTotalD
         public const long LondonBlockNumber = 12_965_000;
         public const long ArrowGlacierBlockNumber = 13_773_000;
         public const long GrayGlacierBlockNumber = 15_050_000;
-        public const ulong ShanghaiBlockTimestamp = ulong.MaxValue - 4;
+        public const ulong ShanghaiBlockTimestamp = ulong.MaxValue - 5;
+        public const ulong ShardingForkBlockTimestamp = ulong.MaxValue - 4;
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Specs/MainNetSpecProvider.cs:62]: QQ: What is the reason with this `-1` or `-5`?

File: src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs [hunk 9]
```diff
@@ -105,5 +109,11 @@ private bool ValidateSignature(Signature? signature, IReleaseSpec spec)
 
             return !spec.ValidateChainId || (signature.V == 27 || signature.V == 28);
         }
+
+        private bool Validate4844Fields(Transaction transaction)
+        {
+            // TODO: Add blobs validation
+            return transaction.Type != TxType.Blob ^ transaction.MaxFeePerDataGas is not null;
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs:116]: NP: Can you invert the condition? `not XOR not` is confusing IMO.

File: src/Nethermind/Nethermind.Specs/ChainSpecStyle/ChainSpec.cs [hunk 10]
```diff
@@ -78,5 +78,7 @@ public class ChainSpec
         public UInt256? TerminalTotalDifficulty { get; set; }
 
         public ulong? ShanghaiTimestamp { get; set; }
+
+        public long? ShardingForkBlockNumber { get; set; }
```
Comment (marcindsobczak) [inline src/Nethermind/Nethermind.Specs/ChainSpecStyle/ChainSpec.cs:82]: It's not used anywhere and `Number` seems not correct anyway

File: src/Nethermind/Nethermind.Blockchain.Test/Validators/TxValidatorTests.cs [hunk 11]
```diff
@@ -215,5 +215,27 @@ public bool MaxFeePerGas_is_required_to_be_greater_than_MaxPriorityFeePerGas(TxT
             TxValidator txValidator = new(TestChainIds.ChainId);
             return txValidator.IsWellFormed(tx, London.Instance);
         }
+
+        [TestCase(TxType.EIP1559, false, ExpectedResult = false)]
```
[the remark this answers is not in the document]
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.Blockchain.Test/Validators/TxValidatorTests.cs:219]: It's false because of `transaction.MaxFeePerGas` < `transaction.MaxPriorityFeePerGas` in `Validate1559GasFields`, looks like in `TransactionBuilder` default value of `MaxFeePerGas` is 0 and `MaxPriorityFeePerGas` is 1. Maybe we need to change it? It's totally not related to this PR so you can fix test by just adding `.WithMaxFeePerGas(1)` when building tx.

[the remark this answers is not in the document]
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.Blockchain.Test/Validators/TxValidatorTests.cs:219]: Or `tx.DecodedMaxFeePerGas = tx.MaxPriorityFeePerGas;` just after building test tx

File: src/Nethermind/Nethermind.Specs/MainNetSpecProvider.cs (hunk 8 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Specs/MainNetSpecProvider.cs:62]: It is may be to show the order of the forks for developers. I did the same just for consistency.

**Review by marcindsobczak — APPROVED** [review]

**Review by smartprogrammer93 — APPROVED** [review]
Looks good about the timestamp activation part, will have to review with you on a call when busy to dive deeper if possible

File: src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs [hunk 12]
```diff
@@ -79,6 +82,7 @@ private bool ValidateChainId(Transaction transaction)
                     return true;
                 case TxType.AccessList:
                 case TxType.EIP1559:
+                case TxType.Blob:
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Validators/TxValidator.cs:85]: should this be just `default` ? Or just have if/else? I don't think we expect this to ever change for new TxTypes.

File: src/Nethermind/Nethermind.Core.Test/Json/TxTypeConverterTests.cs [hunk 13]
```diff
@@ -16,6 +16,7 @@ public class TxTypeConverterTests : ConverterTestBase<TxType>
         [TestCase((TxType)255)]
         [TestCase(TxType.Legacy)]
         [TestCase(TxType.AccessList)]
+        [TestCase(TxType.Blob)]
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core.Test/Json/TxTypeConverterTests.cs:19]: Make it a TestCaseSource and have all TxType by default here + additional values?

File: src/Nethermind/Nethermind.Runner/Nethermind.Runner.csproj [hunk 14]
```diff
@@ -103,6 +103,9 @@
     <None Remove="out\**" />
     <None Remove="keystore\**" />
   </ItemGroup>
+  <ItemGroup>
+    <Content Remove="..\Chains\eip4844_devnet.json" />
+  </ItemGroup>
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Runner/Nethermind.Runner.csproj:108]: ?

File: src/Nethermind/Nethermind.Serialization.Rlp/HeaderDecoder.cs [hunk 15]
```diff
@@ -75,13 +76,20 @@ public class HeaderDecoder : IRlpValueDecoder<BlockHeader>, IRlpStreamDecoder<Bl
             {
                 blockHeader.BaseFeePerGas = decoderContext.DecodeUInt256();
             }
-
+            TODO: fix
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Serialization.Rlp/HeaderDecoder.cs:79]: comment?

File: src/Nethermind/Nethermind.Specs/ChainSpecStyle/ChainSpecBasedSpecProvider.cs [hunk 16]
```diff
@@ -204,6 +204,57 @@ private static ReleaseSpec CreateReleaseSpec(ChainSpec chainSpec, long releaseSt
...
+            releaseSpec.IsEip3607Enabled = (_chainSpec.Parameters.Eip3607Transition ?? long.MaxValue) <= releaseStartBlock;
+            releaseSpec.IsEip4844Enabled = (_chainSpec.Parameters.Eip4844TransitionTimestamp ?? ulong.MaxValue) <= releaseStartTimestamp;
+            releaseSpec.ValidateChainId = (_chainSpec.Parameters.ValidateChainIdTransition ?? 0) <= releaseStartBlock;
+            releaseSpec.ValidateReceipts = ((_chainSpec.Parameters.ValidateReceiptsTransition > 0) ? Math.Max(_chainSpec.Parameters.ValidateReceiptsTransition ?? 0, _chainSpec.Parameters.Eip658Transition ?? 0) : 0) <= releaseStartBlock;
+            releaseSpec.Eip1559FeeCollector = releaseSpec.IsEip1559Enabled && (_chainSpec.Parameters.Eip1559FeeCollectorTransition ?? long.MaxValue) <= releaseStartBlock ? _chainSpec.Parameters.Eip1559FeeCollector : null;
+            releaseSpec.Eip1559BaseFeeMinValue = releaseSpec.IsEip1559Enabled && (_chainSpec.Parameters.Eip1559BaseFeeMinValueTransition ?? long.MaxValue) <= releaseStartBlock ? _chainSpec.Parameters.Eip1559BaseFeeMinValue : null;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Specs/ChainSpecStyle/ChainSpecBasedSpecProvider.cs:257]: why is this changed?

File: src/Nethermind/Nethermind.Specs/Forks/17_Cancun.cs [hunk 17]
```diff
@@ -6,7 +6,7 @@
 
 namespace Nethermind.Specs.Forks
 {
-    public class Cancun : Shanghai
+    public class Cancun : ShardingFork
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Specs/Forks/17_Cancun.cs:9]: Isn't Cancun same as SHardingFork?

File: src/Nethermind/Nethermind.Specs/ReleaseSpec.cs [hunk 18]
```diff
@@ -70,12 +70,14 @@ public ReleaseSpec Clone()
         public bool ValidateReceipts { get; set; }
         public long Eip1559TransitionBlock { get; set; }
         public ulong WithdrawalTimestamp { get; set; }
+        public ulong Eip4844TransitionTimestamp { get; set; } = ulong.MaxValue;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Specs/ReleaseSpec.cs:73]: is ulong.MaxValue needed?

File: src/Nethermind/Nethermind.Blockchain.Test/Validators/TxValidatorTests.cs [hunk 19]
```diff
@@ -243,5 +243,28 @@ public void Transaction_with_init_code_above_max_value_is_rejected_when_eip3860E
             TxValidator txValidator = new(1);
             txValidator.IsWellFormed(tx, releaseSpec).Should().Be(expectedResult);
         }
+
+        [TestCase(TxType.EIP1559, false, ExpectedResult = true)]
+        [TestCase(TxType.Blob, false, ExpectedResult = false)]
+        [TestCase(TxType.EIP1559, true, ExpectedResult = false)]
+        [TestCase(TxType.Blob, true, ExpectedResult = true)]
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain.Test/Validators/TxValidatorTests.cs:250]: You could generate those test with Source for all TxTypes

File: src/Nethermind/Nethermind.Specs/Forks/17_Cancun.cs (hunk 17 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Specs/Forks/17_Cancun.cs:9]: I will remove it in favor to Cancun, but remove transient opcodes from Cancun just in 4844 branch, because I don't think they should be included for now.

File: src/Nethermind/Nethermind.Serialization.Rlp/HeaderDecoder.cs (hunk 15 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Serialization.Rlp/HeaderDecoder.cs]: Intermediate merge state, fixed

File: src/Nethermind/Nethermind.Runner/Nethermind.Runner.csproj (hunk 14 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Runner/Nethermind.Runner.csproj]: Removed

File: src/Nethermind/Nethermind.Core.Test/Json/TxTypeConverterTests.cs (hunk 13 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Core.Test/Json/TxTypeConverterTests.cs]: Made such source, looks quite OK

File: src/Nethermind/Nethermind.Blockchain.Test/Validators/TxValidatorTests.cs (hunk 19 above)
Comment (flcl42) [reply src/Nethermind/Nethermind.Blockchain.Test/Validators/TxValidatorTests.cs:250]: Too complicated to do with sources, the current state is nicely simple to read
